{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {},
   "outputs": [],
   "source": [
    "from utils.fasta_utils import *\n",
    "from utils.proteome_process import *\n",
    "from utils.prefetcher import *\n",
    "from utils.network_utils import *\n",
    "\n",
    "import os\n",
    "from tqdm import tqdm\n",
    "import pandas as pd\n",
    "import hdbscan\n",
    "import faiss\n",
    "from sklearn.metrics.pairwise import cosine_distances,euclidean_distances\n",
    "import numpy as np\n",
    "import networkx as nx\n",
    "\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/home/swapnesh/anaconda3/envs/esmfold/lib/python3.7/site-packages/IPython/core/interactiveshell.py:3552: DtypeWarning: Columns (2,20,29,33,40,57,62,63,73) have mixed types.Specify dtype option on import or set low_memory=False.\n",
      "  exec(code_obj, self.user_global_ns, self.user_ns)\n"
     ]
    }
   ],
   "source": [
    "data_folder = \"/media/microscopie-lcb/swapnesh/protein/embeddings/phages/1Sept2024_INPHARED_db_latest/\"\n",
    "\n",
    "model_name = \"650m\"\n",
    "db_zarr_path = os.path.join(data_folder , f\"ESM2_{model_name}_1Sept24_{model_name}.zarr\")\n",
    "#db_accession_path = os.path.join(data_folder , 'db_accessions.npy')\n",
    "faiss_index_path = os.path.join(data_folder , f\"ESM2_{model_name}_1Sept24_{model_name}.zarrfaiss_index.bin\")\n",
    "eps_values_path = os.path.join(data_folder , f\"ESM2_{model_name}_1Sept24_{model_name}_eps_values_flat_clusters.npy\")\n",
    "#HieVi_tree =  os.path.join(data_folder , \"ESM2_3b_1Sept24_3b.zarr.gexf\")\n",
    "HieVi_INPHARED_ordered_annotation = \"HieVi_INPHARED_ordered_annotation.csv\"\n",
    "\n",
    "\n",
    "#htree = nx.read_gexf(HieVi_tree)\n",
    "db_zarr_store = zarr.open(db_zarr_path,'r')\n",
    "db_accessions = db_zarr_store['accessions'][:]\n",
    "\n",
    "\n",
    "#htree = nx.read_gexf(HieVi_tree)\n",
    "#db_accessions = np.load(db_accession_path)\n",
    "\n",
    "annotations = pd.read_csv(HieVi_INPHARED_ordered_annotation)\n",
    "annotations= annotations[annotations[\"Accession\"].isin(db_accessions)]\n",
    "annotations = annotations.set_index(\"Accession\").loc[db_accessions].reset_index()\n",
    "\n",
    "eps_values = np.load(eps_values_path)\n",
    "index = faiss.read_index(faiss_index_path)\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 57,
   "metadata": {},
   "outputs": [],
   "source": [
    "#filename = \"/media/microscopie-lcb/swapnesh/protein/embeddings/phages/NewGenomes/HieVi_test_case.faa\"\n",
    "filename = \"/home/swapnesh/Downloads/TEST_concatenated.fasta\"\n",
    "filename = \"/home/swapnesh/Downloads/subset.fasta\"\n",
    "\n",
    "# @title Compute phage representations\n",
    "expt_name = \"TEST\"  # @param {type:\"string\"}\n",
    "expt_name = expt_name.replace(' ','_')\n",
    "output_folder = os.path.dirname(filename) + os.sep \n",
    "fasta_path = filename\n",
    "#model_name = \"3b\" # This colab works for 650m only\n",
    "mode = \"mean\"\n",
    "query_zarr_path = os.path.join(output_folder,f\"{expt_name}_{model_name}.zarr\")\n",
    "#!python GenPhageRepresentationsESM2.py {expt_name} {output_folder} {fasta_path} {model_name} {mode}\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 58,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "/home/swapnesh/Downloads/TEST_650m.zarr\n"
     ]
    }
   ],
   "source": [
    "query_zarr_path = f\"{output_folder}{expt_name}_{model_name}.zarr\"\n",
    "print(query_zarr_path)\n",
    "query_zarr_store = zarr.open(query_zarr_path,'r')\n",
    "query_vectors = query_zarr_store['vectors_mean'][:]*1.0\n",
    "query_accessions = query_zarr_store['accessions'][:]\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 59,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "(array([0.00389589, 0.01027566, 0.01244868, 0.01821554, 0.01860557,\n",
       "        0.02378739, 0.02420528, 0.02531965, 0.02565396, 0.02662903,\n",
       "        0.02710264, 0.0286349 ]),\n",
       " (2, 1280))"
      ]
     },
     "execution_count": 59,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "eps_values,query_vectors.shape"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 60,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "([array([  281,   435,   478,   506,   509,   531,   536,   537,   543,\n",
       "           585,   626,   894,   909,   926,  1030,  1133,  1134,  1135,\n",
       "          1185,  1218,  1260,  1478,  1666,  1673,  1717,  1734,  1739,\n",
       "          1752,  1818,  1889,  1890,  1891,  1978,  1988,  2047,  2048,\n",
       "          2049,  2090,  2107,  5001,  5104,  5105,  5106,  5107,  5108,\n",
       "          5109,  5110,  5111,  5112,  5113,  5114,  5115,  5116,  5117,\n",
       "          5118,  5119,  5120,  5121,  5130,  5131,  5132,  5133,  5134,\n",
       "          5135,  5136,  5138,  5139,  5141,  5142,  5143,  5144,  5145,\n",
       "          5154,  5155,  5156,  5158,  5159,  5160,  5161,  5162,  5168,\n",
       "          5208,  5430,  5467,  5784,  5824,  5878,  5895,  5905,  6041,\n",
       "          6153,  6277,  6334,  6628,  6662,  6663,  6664,  6665,  6687,\n",
       "          7321,  7443,  7556,  7648,  7917,  8049,  8096,  8210,  8211,\n",
       "          8215,  8277,  8278,  8279,  8288,  8289,  8362,  8388,  8403,\n",
       "          8409,  8444,  8449,  8911,  9209,  9369,  9376,  9393,  9487,\n",
       "          9503,  9527,  9764,  9765,  9766,  9784,  9990, 10101, 10137,\n",
       "         10162, 10163, 10164, 10169, 10171, 10177, 10187, 10221, 10277,\n",
       "         10278, 11008, 11018, 11020, 11027, 11045, 11106, 11107, 11113,\n",
       "         11227, 11228, 11287, 11356, 11357, 11358, 11359, 11360, 11435,\n",
       "         11637, 11772, 11792, 11938, 11955, 12012, 12054, 12157, 12184,\n",
       "         12408, 12409, 12410, 12412, 12413, 12415, 12416, 12417, 12418,\n",
       "         12419, 12420, 12422, 12423, 12424, 12425, 12426, 12427, 12428,\n",
       "         12429, 12430, 12733, 12734, 12735, 12736, 12737, 12742, 12749,\n",
       "         12761, 12762, 12765, 12768, 12784, 12794, 12814, 12826, 13141,\n",
       "         13391, 13583, 13584, 13585, 13586, 13595, 13938, 13953, 13954,\n",
       "         14052, 14056, 14107, 14111, 14113, 14240, 14281, 14283, 14451,\n",
       "         14456, 14507, 14531, 14537, 14564, 14621, 14622, 14784, 14788,\n",
       "         15006, 15051, 15052, 15066, 15138, 15191, 15289, 15292, 15315,\n",
       "         15316, 15454, 15455, 15457, 15458, 15459, 15467, 15512, 15626,\n",
       "         15627, 15628, 15629, 15630, 15631, 16228, 16229, 16232, 16238,\n",
       "         16239, 16240, 16247, 16249, 16250, 16251, 16252, 16253, 16254,\n",
       "         16784, 16785, 16786, 16816, 16832, 17252, 17980, 18129, 18130,\n",
       "         18131, 18464, 18566, 18634, 19251, 19295, 19346, 19451, 19472,\n",
       "         19561, 19935, 19995, 20470, 20472, 20473, 20474, 20475, 20493,\n",
       "         20895, 20918, 21020, 21021, 21075, 21076, 21248, 21262, 21353,\n",
       "         21476, 21477, 21478, 21738, 21856, 21977, 22041, 22052, 22053,\n",
       "         22071, 22098, 22099, 22153, 22161, 22165, 22167, 22321, 22372,\n",
       "         22420, 22515, 22906, 22907, 22908, 22909, 22910, 23016, 23472,\n",
       "         23473, 23479, 23481, 23517, 24127, 24131, 24272, 24277, 24278,\n",
       "         24279, 24280, 24281, 24282, 24315, 24317, 24343, 24344, 24345,\n",
       "         24347]),\n",
       "  array([  281,   435,   470,   478,   506,   509,   531,   536,   537,\n",
       "           543,   585,   626,   723,   740,   894,   909,   926,  1030,\n",
       "          1133,  1134,  1135,  1185,  1218,  1260,  1478,  1666,  1673,\n",
       "          1717,  1729,  1734,  1739,  1752,  1799,  1818,  1830,  1889,\n",
       "          1890,  1891,  1978,  1988,  2012,  2047,  2048,  2049,  2090,\n",
       "          2107,  5001,  5104,  5105,  5106,  5107,  5108,  5109,  5110,\n",
       "          5111,  5112,  5113,  5114,  5115,  5116,  5117,  5118,  5119,\n",
       "          5120,  5121,  5130,  5131,  5132,  5133,  5134,  5135,  5136,\n",
       "          5138,  5139,  5141,  5142,  5143,  5144,  5145,  5154,  5155,\n",
       "          5156,  5158,  5159,  5160,  5161,  5162,  5168,  5208,  5430,\n",
       "          5467,  5488,  5507,  5784,  5824,  5878,  5895,  5905,  6041,\n",
       "          6153,  6277,  6334,  6337,  6628,  6662,  6663,  6664,  6665,\n",
       "          6687,  6797,  6953,  7321,  7392,  7436,  7443,  7555,  7556,\n",
       "          7635,  7648,  7917,  8027,  8030,  8049,  8096,  8210,  8211,\n",
       "          8215,  8270,  8277,  8278,  8279,  8288,  8289,  8362,  8388,\n",
       "          8403,  8409,  8443,  8444,  8449,  8911,  9209,  9369,  9376,\n",
       "          9393,  9487,  9503,  9527,  9764,  9765,  9766,  9784,  9990,\n",
       "         10101, 10136, 10162, 10163, 10164, 10169, 10171, 10177, 10187,\n",
       "         10221, 10277, 10278, 10292, 11008, 11018, 11020, 11027, 11045,\n",
       "         11106, 11107, 11113, 11227, 11228, 11287, 11356, 11357, 11358,\n",
       "         11359, 11360, 11435, 11603, 11637, 11772, 11792, 11807, 11938,\n",
       "         11955, 12012, 12054, 12145, 12157, 12184, 12408, 12409, 12410,\n",
       "         12412, 12413, 12415, 12416, 12417, 12418, 12419, 12420, 12422,\n",
       "         12423, 12424, 12425, 12426, 12427, 12428, 12429, 12430, 12733,\n",
       "         12734, 12735, 12736, 12737, 12742, 12749, 12761, 12762, 12763,\n",
       "         12765, 12768, 12784, 12794, 12814, 12826, 12848, 12884, 12924,\n",
       "         12928, 12965, 12971, 12972, 12990, 13141, 13391, 13583, 13584,\n",
       "         13585, 13586, 13595, 13938, 13949, 13950, 13951, 13952, 13953,\n",
       "         13954, 14052, 14056, 14107, 14111, 14113, 14240, 14281, 14283,\n",
       "         14451, 14456, 14494, 14507, 14531, 14537, 14564, 14621, 14622,\n",
       "         14784, 14785, 14788, 15006, 15051, 15052, 15066, 15138, 15159,\n",
       "         15191, 15289, 15292, 15315, 15316, 15401, 15402, 15454, 15455,\n",
       "         15457, 15458, 15459, 15467, 15512, 15562, 15626, 15627, 15628,\n",
       "         15629, 15630, 15631, 15699, 16228, 16229, 16232, 16238, 16239,\n",
       "         16240, 16247, 16249, 16250, 16251, 16252, 16253, 16254, 16670,\n",
       "         16784, 16785, 16786, 16816, 16832, 16935, 17252, 17980, 18129,\n",
       "         18130, 18131, 18464, 18566, 18634, 18764, 19251, 19295, 19346,\n",
       "         19437, 19451, 19472, 19561, 19823, 19935, 19936, 19937, 19995,\n",
       "         20287, 20470, 20472, 20473, 20474, 20475, 20493, 20895, 20918,\n",
       "         21020, 21021, 21075, 21076, 21248, 21262, 21353, 21476, 21477,\n",
       "         21478, 21738, 21856, 21977, 22040, 22041, 22052, 22053, 22054,\n",
       "         22071, 22098, 22099, 22153, 22161, 22165, 22167, 22256, 22321,\n",
       "         22372, 22420, 22515, 22906, 22907, 22908, 22909, 22910, 23016,\n",
       "         23472, 23473, 23479, 23481, 23517, 24127, 24131, 24272, 24277,\n",
       "         24278, 24279, 24280, 24281, 24282, 24285, 24315, 24317, 24343,\n",
       "         24344, 24345, 24347, 24349])],\n",
       " [array([0.000418  , 0.00099336, 0.0003909 , 0.0015572 , 0.00029016,\n",
       "         0.00149538, 0.00020794, 0.00144884, 0.00045931, 0.00014273,\n",
       "         0.00108971, 0.00150691, 0.00111442, 0.0002099 , 0.0005826 ,\n",
       "         0.00093568, 0.0009615 , 0.00093545, 0.00108299, 0.00027418,\n",
       "         0.0012043 , 0.00100132, 0.00021048, 0.00092855, 0.00144783,\n",
       "         0.00100886, 0.00152596, 0.00148221, 0.00117479, 0.00118203,\n",
       "         0.00158308, 0.00104321, 0.00091816, 0.00044502, 0.00106081,\n",
       "         0.0011525 , 0.0011305 , 0.        , 0.00080727, 0.00020339,\n",
       "         0.00108466, 0.00110787, 0.00108745, 0.00109127, 0.0011082 ,\n",
       "         0.00115064, 0.00115025, 0.00114563, 0.00115064, 0.00115064,\n",
       "         0.00115064, 0.00115064, 0.00119053, 0.00115064, 0.00115064,\n",
       "         0.00115521, 0.00114894, 0.00115064, 0.00110701, 0.00109096,\n",
       "         0.00115187, 0.00123319, 0.00115124, 0.00114792, 0.00106306,\n",
       "         0.00107296, 0.00107888, 0.00115883, 0.00114349, 0.00108745,\n",
       "         0.00108452, 0.00107269, 0.00136197, 0.00105092, 0.00120173,\n",
       "         0.00104581, 0.00020203, 0.0010466 , 0.00115073, 0.00115057,\n",
       "         0.00020682, 0.00121518, 0.00129122, 0.00156674, 0.00018154,\n",
       "         0.00055115, 0.00137572, 0.0011005 , 0.00101432, 0.00093784,\n",
       "         0.00036706, 0.00148605, 0.0009145 , 0.00087635, 0.00042727,\n",
       "         0.00030157, 0.0004532 , 0.00018409, 0.0009676 , 0.0009887 ,\n",
       "         0.00116515, 0.00136709, 0.00067516, 0.00030355, 0.00110161,\n",
       "         0.00098454, 0.0011204 , 0.00140937, 0.00044542, 0.00102968,\n",
       "         0.00095903, 0.00099286, 0.00035237, 0.00029699, 0.00039673,\n",
       "         0.00094397, 0.00020101, 0.00109316, 0.00109725, 0.0010292 ,\n",
       "         0.00144279, 0.00152323, 0.00132936, 0.00101809, 0.00110464,\n",
       "         0.00110805, 0.0010625 , 0.00104236, 0.00012437, 0.00012561,\n",
       "         0.00015295, 0.00082216, 0.00116381, 0.00114097, 0.00148242,\n",
       "         0.00111579, 0.0010275 , 0.00096654, 0.00039496, 0.00025529,\n",
       "         0.00027629, 0.00027514, 0.0001691 , 0.0010384 , 0.00081445,\n",
       "         0.00016343, 0.0001967 , 0.00042498, 0.00054997, 0.00013957,\n",
       "         0.0003871 , 0.00028013, 0.00028348, 0.00097347, 0.0010589 ,\n",
       "         0.00157249, 0.00097902, 0.00090736, 0.00091043, 0.00090915,\n",
       "         0.00097511, 0.00151234, 0.00121624, 0.00091295, 0.00139326,\n",
       "         0.00023504, 0.00048223, 0.00096081, 0.00126916, 0.00102032,\n",
       "         0.00012067, 0.00111263, 0.00134022, 0.00116591, 0.00116758,\n",
       "         0.00117533, 0.00117946, 0.00128277, 0.00086912, 0.00088516,\n",
       "         0.00125227, 0.00119288, 0.00112709, 0.00117573, 0.00132674,\n",
       "         0.00118313, 0.00131399, 0.00134757, 0.00126513, 0.00134977,\n",
       "         0.00116806, 0.00127683, 0.00106396, 0.0008747 , 0.00086704,\n",
       "         0.00097373, 0.00091941, 0.00102332, 0.00118429, 0.00102124,\n",
       "         0.00055281, 0.00029302, 0.00059561, 0.00097443, 0.00029283,\n",
       "         0.00021639, 0.000918  , 0.00146271, 0.00050811, 0.00045834,\n",
       "         0.00046091, 0.00050971, 0.00139416, 0.00049591, 0.00155005,\n",
       "         0.00151059, 0.00091794, 0.0009115 , 0.00042579, 0.00097943,\n",
       "         0.00103744, 0.00020852, 0.00082398, 0.00074395, 0.00104517,\n",
       "         0.00048716, 0.00153131, 0.00109803, 0.0013737 , 0.00137611,\n",
       "         0.00102518, 0.00149044, 0.00140744, 0.00091816, 0.00096922,\n",
       "         0.00053083, 0.00033056, 0.00107597, 0.00102005, 0.00099759,\n",
       "         0.00137504, 0.00149285, 0.00114788, 0.00107617, 0.00103652,\n",
       "         0.00103652, 0.0011634 , 0.00094305, 0.00103652, 0.00103269,\n",
       "         0.00088871, 0.00103075, 0.00103652, 0.00103652, 0.00102586,\n",
       "         0.0010826 , 0.00100447, 0.00111219, 0.00111192, 0.00114256,\n",
       "         0.00106712, 0.00106758, 0.00117112, 0.00117766, 0.00124154,\n",
       "         0.00111192, 0.00133256, 0.00122152, 0.00110795, 0.0010672 ,\n",
       "         0.00031392, 0.00078697, 0.00038252, 0.00092105, 0.00129386,\n",
       "         0.0003409 , 0.00089032, 0.00116877, 0.00121744, 0.00088744,\n",
       "         0.00044127, 0.00121266, 0.0001113 , 0.00103351, 0.00053506,\n",
       "         0.00103839, 0.00111849, 0.00106181, 0.00096149, 0.00123969,\n",
       "         0.00082667, 0.00015843, 0.00017277, 0.00029977, 0.00023749,\n",
       "         0.00022471, 0.00098842, 0.00105127, 0.00152003, 0.00097696,\n",
       "         0.00103283, 0.0004495 , 0.00052712, 0.00028198, 0.00103855,\n",
       "         0.00101921, 0.00107867, 0.00109975, 0.00104848, 0.00120529,\n",
       "         0.00048328, 0.0011988 , 0.00151545, 0.00152268, 0.00143776,\n",
       "         0.00110665, 0.00108957, 0.00123152, 0.00094416, 0.00134235,\n",
       "         0.00096001, 0.00097474, 0.00120956, 0.0002266 , 0.00039322,\n",
       "         0.00093258, 0.00121646, 0.00159016, 0.00125431, 0.00157959,\n",
       "         0.00127962, 0.00022313, 0.00118761, 0.00120831, 0.00108193,\n",
       "         0.0011224 , 0.00085167, 0.00012752, 0.00076072, 0.00104154,\n",
       "         0.00019887, 0.00015082, 0.00020011, 0.00023478, 0.00016488,\n",
       "         0.00044458, 0.00136483, 0.0005604 , 0.0009431 , 0.00145436,\n",
       "         0.00113239, 0.00115978], dtype=float32),\n",
       "  array([2.97638820e-04, 9.00353538e-04, 1.54852215e-03, 2.95395264e-04,\n",
       "         1.26851758e-03, 2.19333015e-04, 1.15783559e-03, 1.38997013e-04,\n",
       "         1.11387670e-03, 3.37051577e-04, 1.48354433e-04, 9.70765483e-04,\n",
       "         1.54530234e-03, 1.55479764e-03, 1.13293226e-03, 1.03841559e-03,\n",
       "         1.42026198e-04, 4.04098304e-04, 9.16447723e-04, 9.49630397e-04,\n",
       "         9.16367746e-04, 9.19032667e-04, 1.29565800e-04, 9.35309974e-04,\n",
       "         9.59353056e-04, 1.58390525e-04, 7.56835216e-04, 1.30135124e-03,\n",
       "         1.46611454e-03, 9.42574232e-04, 1.45216123e-03, 1.17733236e-03,\n",
       "         1.54962332e-03, 1.28735253e-03, 1.49700779e-03, 9.56987147e-04,\n",
       "         1.34105771e-03, 9.44322965e-04, 8.53357720e-04, 2.02664058e-04,\n",
       "         1.44112692e-03, 9.23533109e-04, 9.87117179e-04, 9.71293659e-04,\n",
       "         1.69101811e-04, 7.69550912e-04, 1.24292143e-04, 1.06763700e-03,\n",
       "         1.10967772e-03, 1.07280281e-03, 1.08106667e-03, 1.17455574e-03,\n",
       "         1.10207556e-03, 1.10177509e-03, 1.09754398e-03, 1.10207556e-03,\n",
       "         1.10207556e-03, 1.10207556e-03, 1.10207556e-03, 1.16116134e-03,\n",
       "         1.10207556e-03, 1.10207556e-03, 1.12434174e-03, 1.10075413e-03,\n",
       "         1.10207556e-03, 1.17353955e-03, 1.08106330e-03, 1.10520120e-03,\n",
       "         1.19884592e-03, 1.10256346e-03, 1.09887472e-03, 1.04863639e-03,\n",
       "         1.13615044e-03, 1.06524304e-03, 1.13259931e-03, 1.09526317e-03,\n",
       "         1.07280281e-03, 1.06939115e-03, 1.00132287e-03, 1.35987066e-03,\n",
       "         1.11278286e-03, 1.13136251e-03, 1.00167887e-03, 1.18618351e-04,\n",
       "         1.03305117e-03, 1.12149504e-03, 1.10217219e-03, 1.28238578e-04,\n",
       "         1.07586640e-03, 1.05139974e-03, 1.27710495e-03, 1.52136490e-03,\n",
       "         1.38982921e-03, 1.06284948e-04, 4.26653016e-04, 1.18344626e-03,\n",
       "         9.11493320e-04, 9.34798736e-04, 8.36979249e-04, 2.32348379e-04,\n",
       "         1.44907343e-03, 9.03287379e-04, 1.42845290e-03, 8.45900329e-04,\n",
       "         3.07751325e-04, 3.79673380e-04, 2.93541874e-04, 1.98205584e-04,\n",
       "         8.99121398e-04, 1.48446742e-03, 1.45491469e-03, 8.97138787e-04,\n",
       "         1.54293352e-03, 1.40229310e-03, 1.05727860e-03, 1.31835672e-03,\n",
       "         1.08851376e-03, 1.51459977e-03, 6.54433738e-04, 3.26707755e-04,\n",
       "         1.59617804e-03, 1.50216883e-03, 9.89588443e-04, 1.02468091e-03,\n",
       "         1.02185761e-03, 1.12493127e-03, 3.82130034e-04, 1.36547221e-03,\n",
       "         1.08736428e-03, 1.00676343e-03, 1.03806588e-03, 2.48289492e-04,\n",
       "         1.81569252e-04, 2.43893737e-04, 9.85283637e-04, 1.37834984e-04,\n",
       "         1.13563379e-03, 1.37551536e-03, 1.02006958e-03, 9.21399216e-04,\n",
       "         1.58353266e-03, 1.59302284e-03, 1.09012797e-03, 9.38640791e-04,\n",
       "         1.04075042e-03, 9.61116108e-04, 8.12804035e-04, 1.07424543e-03,\n",
       "         1.43427635e-04, 1.44713966e-04, 1.44478763e-04, 9.23421932e-04,\n",
       "         1.13117159e-03, 1.19903148e-03, 1.56850799e-03, 1.05489499e-03,\n",
       "         9.31550050e-04, 9.05985013e-04, 2.94642989e-04, 2.23671872e-04,\n",
       "         1.24714614e-04, 1.53396366e-04, 0.00000000e+00, 1.10061746e-03,\n",
       "         8.00991082e-04, 1.56608713e-03, 1.53458139e-04, 1.35278475e-04,\n",
       "         2.34369509e-04, 4.22090001e-04, 1.73561231e-04, 2.50122626e-04,\n",
       "         1.76231202e-04, 1.90322608e-04, 9.02311702e-04, 9.17100464e-04,\n",
       "         1.38842897e-03, 1.02904765e-03, 8.96910788e-04, 8.99866689e-04,\n",
       "         8.99937702e-04, 9.63451923e-04, 1.22272782e-03, 1.43753877e-03,\n",
       "         9.80861834e-04, 8.22507543e-04, 1.32307655e-03, 1.40732073e-03,\n",
       "         1.12767593e-04, 3.59061261e-04, 8.94865370e-04, 1.07135379e-03,\n",
       "         1.41790393e-03, 1.02284807e-03, 1.03758794e-04, 8.98739556e-04,\n",
       "         1.07011909e-03, 9.30822105e-04, 9.32135270e-04, 9.40878177e-04,\n",
       "         9.42844839e-04, 1.38068502e-03, 7.82107993e-04, 8.02134280e-04,\n",
       "         9.91588342e-04, 9.51002468e-04, 9.16007382e-04, 9.41254431e-04,\n",
       "         1.04974676e-03, 9.45914187e-04, 1.03568938e-03, 1.06511381e-03,\n",
       "         1.00260903e-03, 1.07435137e-03, 9.32639115e-04, 1.10736606e-03,\n",
       "         1.02331233e-03, 8.83527100e-04, 9.01551160e-04, 8.87018978e-04,\n",
       "         9.57931625e-04, 1.01200619e-03, 1.23502139e-03, 1.10293995e-03,\n",
       "         1.32100703e-03, 3.55948636e-04, 1.49235973e-04, 3.54969059e-04,\n",
       "         9.18817939e-04, 1.89578510e-04, 1.46199163e-04, 1.59796607e-03,\n",
       "         1.44211412e-03, 1.43454620e-03, 1.43454620e-03, 1.43454620e-03,\n",
       "         1.43846578e-03, 1.48308428e-03, 1.43846578e-03, 9.56328935e-04,\n",
       "         1.25831494e-03, 2.90643802e-04, 2.28855701e-04, 2.31443366e-04,\n",
       "         2.91466218e-04, 1.09761022e-03, 2.90892902e-04, 1.36693439e-03,\n",
       "         1.40666729e-03, 1.47575152e-03, 1.42041524e-03, 1.36807188e-03,\n",
       "         1.28266728e-03, 9.13783966e-04, 1.01390434e-03, 3.11399752e-04,\n",
       "         9.30305570e-04, 1.03225885e-03, 1.14833994e-04, 6.15609926e-04,\n",
       "         5.25870477e-04, 9.36177210e-04, 3.30090406e-04, 1.50816259e-03,\n",
       "         1.27460342e-03, 1.00836181e-03, 1.19093980e-03, 1.24672172e-03,\n",
       "         1.01611495e-03, 1.18557969e-03, 1.13228988e-03, 1.56850088e-03,\n",
       "         8.53357720e-04, 8.65774695e-04, 3.62451887e-04, 2.23774026e-04,\n",
       "         1.07734883e-03, 1.01791951e-03, 1.33759412e-03, 9.23917047e-04,\n",
       "         1.41723477e-03, 1.21135800e-03, 9.64110543e-04, 1.11979165e-03,\n",
       "         1.38196873e-03, 1.39938411e-03, 9.31726187e-04, 9.31726187e-04,\n",
       "         1.01942557e-03, 8.73831683e-04, 9.31726187e-04, 9.30249807e-04,\n",
       "         6.67725573e-04, 1.47995795e-03, 9.26906243e-04, 9.31726187e-04,\n",
       "         9.31726187e-04, 9.24215419e-04, 9.87110427e-04, 9.05229826e-04,\n",
       "         1.50036695e-03, 1.10437209e-03, 1.10380840e-03, 1.12007628e-03,\n",
       "         1.06684444e-03, 1.06717227e-03, 1.20079995e-03, 1.20191311e-03,\n",
       "         1.28341932e-03, 1.10380840e-03, 1.42747979e-03, 1.27364998e-03,\n",
       "         1.11679907e-03, 1.06713152e-03, 1.51688035e-03, 2.16183267e-04,\n",
       "         5.51438425e-04, 1.98089416e-04, 8.11859849e-04, 1.11167505e-03,\n",
       "         1.49319205e-03, 2.00896611e-04, 9.15527518e-04, 1.23883085e-03,\n",
       "         9.98160103e-04, 9.39460355e-04, 3.03650158e-04, 1.10336137e-03,\n",
       "         1.19569726e-04, 1.45261374e-03, 9.40653146e-04, 6.20123988e-04,\n",
       "         9.45591892e-04, 1.57980074e-03, 9.44914646e-04, 1.09480415e-03,\n",
       "         1.05986744e-03, 1.43642537e-03, 1.04446989e-03, 1.45485241e-03,\n",
       "         1.55256013e-03, 8.33189581e-04, 1.44121028e-03, 7.01124736e-05,\n",
       "         9.29562520e-05, 1.90578285e-04, 1.23457052e-04, 1.17079740e-04,\n",
       "         1.00005546e-03, 1.06084754e-03, 1.19922194e-03, 1.01661356e-03,\n",
       "         1.04494591e-03, 3.09001160e-04, 3.72605515e-04, 1.93686312e-04,\n",
       "         1.11314259e-03, 9.98531235e-04, 1.07468641e-03, 1.17537752e-03,\n",
       "         1.09974353e-03, 1.05622993e-03, 3.18655977e-04, 1.15402974e-03,\n",
       "         1.51070207e-03, 1.42658479e-03, 1.35757460e-03, 1.33104902e-03,\n",
       "         1.51164492e-03, 9.69746674e-04, 1.00526866e-03, 9.91842942e-04,\n",
       "         8.25729978e-04, 1.08733959e-03, 8.94310535e-04, 8.93813907e-04,\n",
       "         1.49251474e-03, 1.04742404e-03, 2.47107324e-04, 2.69865937e-04,\n",
       "         8.57670384e-04, 1.16015016e-03, 1.31280895e-03, 1.19954639e-03,\n",
       "         1.33214495e-03, 1.14881352e-03, 1.77812486e-04, 1.19161443e-03,\n",
       "         1.22212607e-03, 1.02029764e-03, 1.05493492e-03, 1.01454579e-03,\n",
       "         7.48064922e-05, 7.45210156e-04, 1.13023492e-03, 8.09765988e-05,\n",
       "         1.71279447e-04, 1.50096472e-04, 1.89724989e-04, 1.21516336e-04,\n",
       "         3.05319845e-04, 1.48308626e-03, 1.22199825e-03, 3.76626995e-04,\n",
       "         9.35079646e-04, 1.25235273e-03, 9.84365121e-04, 9.81282443e-04,\n",
       "         1.32713362e-03], dtype=float32)])"
      ]
     },
     "execution_count": 60,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "radius = 0.0016\n",
    "\n",
    "# Perform range search\n",
    "lims, distances, indices = index.range_search(query_vectors, radius)\n",
    "indices = [indices[lims[i]:lims[i+1]] for i in range(len(query_vectors)) ]\n",
    "distances = [distances[lims[i]:lims[i+1]] for i in range(len(query_vectors)) ]\n",
    "indices,distances"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 61,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "(array([[0.],\n",
       "        [0.]], dtype=float32),\n",
       " array([[ 2090],\n",
       "        [10221]]))"
      ]
     },
     "execution_count": 61,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "index.search(query_vectors, 1)\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 43,
   "metadata": {},
   "outputs": [
    {
     "ename": "NameError",
     "evalue": "name 'htree' is not defined",
     "output_type": "error",
     "traceback": [
      "\u001b[0;31m---------------------------------------------------------------------------\u001b[0m",
      "\u001b[0;31mNameError\u001b[0m                                 Traceback (most recent call last)",
      "\u001b[0;32m/tmp/ipykernel_642016/4273404259.py\u001b[0m in \u001b[0;36m<module>\u001b[0;34m\u001b[0m\n\u001b[1;32m     21\u001b[0m \u001b[0mnew_accession\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0;34m[\u001b[0m\u001b[0;34m]\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m     22\u001b[0m \u001b[0;32mfor\u001b[0m \u001b[0ma\u001b[0m \u001b[0;32min\u001b[0m \u001b[0mall_nearest_accessions\u001b[0m\u001b[0;34m:\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0;32m---> 23\u001b[0;31m     \u001b[0mnew_accession\u001b[0m \u001b[0;34m+=\u001b[0m \u001b[0mfind_predecessor_and_leaves\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mhtree\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0ma\u001b[0m\u001b[0;34m,\u001b[0m\u001b[0mdistance_in_tree\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0m\u001b[1;32m     24\u001b[0m \u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m     25\u001b[0m \u001b[0;34m\u001b[0m\u001b[0m\n",
      "\u001b[0;31mNameError\u001b[0m: name 'htree' is not defined"
     ]
    }
   ],
   "source": [
    "hievi_cluster_prefix = 'HC_p'\n",
    "k_neighbours = 16 # \n",
    "distance_threshold = 0.023 # \n",
    "distance_in_tree = 2\n",
    "\n",
    "distances, indices = index.search(query_vectors, k_neighbours)\n",
    "valid_idx = distances[:,0] < distance_threshold\n",
    "invalid_idx = np.logical_not(valid_idx)\n",
    "\n",
    "if len(np.where(invalid_idx)[0]):\n",
    "    print('Cannot classifiy: ' ,len(np.where(invalid_idx)[0]))\n",
    "    invalid_query_df = pd.DataFrame({\"Accession\": query_accessions[invalid_idx]})\n",
    "    invalid_query_df.to_csv(query_zarr_path[:-5] + \"_HieVi_Unclassifieds.csv\")\n",
    "\n",
    "D1 = np.squeeze(distances[valid_idx,0])\n",
    "I1 = np.squeeze(indices[valid_idx,0])\n",
    "\n",
    "all_indices = np.unique(np.array(I1))\n",
    "all_nearest_accessions = db_accessions[all_indices]\n",
    "\n",
    "new_accession = []\n",
    "for a in all_nearest_accessions:\n",
    "    new_accession += find_predecessor_and_leaves(htree, a,distance_in_tree)\n",
    "    \n",
    "\n",
    "all_nearest_accessions = np.unique(np.ravel(np.array(new_accession)))\n",
    "all_indices = np.array([np.where(db_accessions==a)[0] for a in all_nearest_accessions])\n",
    "\n",
    "subset_db_vectors = np.array([index.reconstruct(int(a)) for a in all_indices])\n",
    "subset_db_vectors.shape\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 48,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Loaded 102 Accessions from database.\n"
     ]
    }
   ],
   "source": [
    "hievi_cluster_prefix = 'HC_'\n",
    "k_neighbours = 8 # \n",
    "distance_threshold = 0.023 # \n",
    "distance_in_tree = 2\n",
    "\n",
    "distances, indices = index.search(query_vectors, k_neighbours)\n",
    "valid_idx = distances[:,0] < distance_threshold\n",
    "invalid_idx = np.logical_not(valid_idx)\n",
    "\n",
    "if len(np.where(invalid_idx)[0]):\n",
    "    print('Cannot classifiy: ' ,len(np.where(invalid_idx)[0]))\n",
    "    invalid_query_df = pd.DataFrame({\"Accession\": query_accessions[invalid_idx]})\n",
    "    invalid_query_df.to_csv(query_zarr_path[:-5] + \"_HieVi_Unclassifieds.csv\")\n",
    "\n",
    "all_indices = np.unique(np.ravel(np.array(indices[valid_idx])))\n",
    "subset_db_vectors = np.array([db_zarr_store['vectors_mean'][i] for i in all_indices])\n",
    "distances, indices = index.search(subset_db_vectors, k_neighbours)\n",
    "\n",
    "\n",
    "D1 = np.squeeze(distances)\n",
    "I1 = np.squeeze(indices)\n",
    "\n",
    "all_indices = np.unique(np.array(I1))\n",
    "all_nearest_accessions = db_accessions[all_indices]\n",
    "\n",
    "#subset_db_vectors = db_zarr_store['vectors_mean'][all_indices,:]\n",
    "\n",
    "subset_db_vectors = np.array([db_zarr_store['vectors_mean'][i] for i in all_indices])\n",
    "\n",
    "print(f\"Loaded {subset_db_vectors.shape[0]} Accessions from database.\" )"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 49,
   "metadata": {},
   "outputs": [],
   "source": [
    "# @title Combine query and nearest phages from database data\n",
    "\n",
    "# Get nearest accessions in tree\n",
    "nearest_accessions = annotations[annotations[\"Accession\"].isin(all_nearest_accessions)]\n",
    "nearest_accessions = nearest_accessions.set_index(\"Accession\").loc[all_nearest_accessions].reset_index()\n",
    "\n",
    "# Combine query and database data\n",
    "query_df = pd.DataFrame({\"Accession\": query_accessions[valid_idx],\"Query\": \"yes\"})\n",
    "annotation_df = pd.concat([nearest_accessions, query_df], axis=0)\n",
    "mprs = np.concatenate((subset_db_vectors, query_vectors[valid_idx]), axis=0)\n",
    "\n",
    "# Perform clustering\n",
    "# dist_scaled = euclidean_distances(mprs).astype(\"double\")\n",
    "# clusterer = hdbscan.HDBSCAN(\n",
    "#     min_cluster_size=2,\n",
    "#     n_jobs=32,\n",
    "#     min_samples=1,\n",
    "#     allow_single_cluster=False,\n",
    "#     cluster_selection_method=\"leaf\",\n",
    "#     metric=\"precomputed\",\n",
    "#     gen_min_span_tree=True\n",
    "# )\n",
    "# clusterer.fit(dist_scaled)\n",
    "\n",
    "clusterer = hdbscan.HDBSCAN(\n",
    "    min_cluster_size=2,\n",
    "    min_samples=1,\n",
    "    allow_single_cluster=False,\n",
    "    cluster_selection_method=\"leaf\",\n",
    "    metric=\"euclidean\",\n",
    "    gen_min_span_tree=True\n",
    ")\n",
    "clusterer.fit(mprs)\n",
    "\n",
    "annotation_df[\"HieVi_cluster\"] = clusterer.labels_\n",
    "for i,eps in enumerate(eps_values):\n",
    "    annotation_df[hievi_cluster_prefix+str(i)] = clusterer.dbscan_clustering(cut_distance=eps,min_cluster_size=2)\n",
    "#annotation_df\n",
    "\n",
    "import re\n",
    "# Function to clean text\n",
    "def clean_text(text):\n",
    "    text = re.sub(r'[^\\w\\s-]', '', text)  # Allow dashes by including \"-\" in the character class\n",
    "    text = re.sub(r'%', '', text)  # Remove percent signs\n",
    "    text = text.strip()  # Remove leading/trailing spaces\n",
    "    return text    \n",
    "min_lambda = 1/0.023 # @param {type:\"slider\", min:-1, max:32, step:1}\n",
    "\n",
    "node_attributes = ['Accession',\"Query\", 'Virus_Description', 'Realm', 'Kingdom', 'Phylum',\n",
    "       'Class', 'Order', 'Family', 'Subfamily', 'Genus','Host_Enveloppe',\n",
    "       'Host_Isolation', 'Host_species', 'Host_order',\n",
    "       'Host_phylum', 'Molecule_type','HieVi_cluster', 'HC_0', 'HC_1',\n",
    "       'HC_2', 'HC_3', 'HC_4', 'HC_5', 'HC_6', 'HC_7', 'HC_8', 'HC_9', 'HC_10',\n",
    "       'HC_11']\n",
    "\n",
    "df = annotation_df[node_attributes].copy()\n",
    "# Apply cleaning to column names and all values\n",
    "df.columns = [clean_text(col) for col in df.columns]\n",
    "df = df.applymap(lambda x: clean_text(str(x)))  # Apply to all values\n",
    "\n",
    "\n",
    "min_lambda =13.00 # @param {type:\"slider\", min:-1, max:32, step:1}\n",
    "# Create and save network\n",
    "G = make_network(clusterer, df,min_lambda=min_lambda)\n",
    "nx.write_gexf(G, query_zarr_path[:-5] + \"_HieVi.gexf\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 50,
   "metadata": {},
   "outputs": [],
   "source": [
    "\n",
    "import networkx as nx\n",
    "import plotly.graph_objects as go\n",
    "import pydot\n",
    "import random\n",
    "import plotly.colors as pc\n",
    "from networkx.drawing.nx_pydot import graphviz_layout\n",
    "\n",
    "def plot_hierarchical_graph(G,layout = \"radial\"):\n",
    "    if not isinstance(G, nx.DiGraph):\n",
    "        raise ValueError(\"Graph must be a directed graph (nx.DiGraph)\")\n",
    "\n",
    "    # Ensure all node names are strings\n",
    "    G = nx.relabel_nodes(G, {node: str(node) for node in G.nodes()})\n",
    "\n",
    "    # Fix attributes: Rename \"name\" -> \"name_old\" if it exists\n",
    "    for node in G.nodes():\n",
    "        if \"name\" in G.nodes[node]:\n",
    "            G.nodes[node][\"name_old\"] = G.nodes[node].pop(\"name\")\n",
    "\n",
    "    # Remove problematic attributes (lists, dicts, tuples)\n",
    "    for node in G.nodes():\n",
    "        for key, value in list(G.nodes[node].items()):\n",
    "            if isinstance(value, (list, dict, tuple)):\n",
    "                del G.nodes[node][key]\n",
    "\n",
    "    # Convert all attributes to strings\n",
    "    for node in G.nodes():\n",
    "        for key in G.nodes[node]:\n",
    "            G.nodes[node][key] = str(G.nodes[node][key])\n",
    "\n",
    "    try:\n",
    "        if layout ==\"radial\":\n",
    "            pos = graphviz_layout(G, prog=\"twopi\")\n",
    "        else:\n",
    "            pos = graphviz_layout(G, prog=\"dot\")\n",
    "    except Exception as e:\n",
    "        print(\"Error: Graphviz failed!\", e)\n",
    "        return\n",
    "\n",
    "    # Get unique genera and assign random colors\n",
    "    unique_genera = set(G.nodes[node].get(\"Genus\", \"Unclassified\") for node in G.nodes())\n",
    "    color_map = {genus: random.choice(pc.sequential.Electric) for genus in unique_genera} #DEFAULT_PLOTLY_COLORS\n",
    "\n",
    "    # Edge positions\n",
    "    edge_x, edge_y = [], []\n",
    "    for edge in G.edges():\n",
    "        x0, y0 = pos[edge[0]]\n",
    "        x1, y1 = pos[edge[1]]\n",
    "        edge_x += [x0, x1, None]\n",
    "        edge_y += [y0, y1, None]\n",
    "\n",
    "    edge_trace = go.Scatter(\n",
    "        x=edge_x, y=edge_y, line=dict(width=1, color='black'),\n",
    "        hoverinfo='none', mode='lines'\n",
    "    )\n",
    "\n",
    "    # Node positions, colors, tooltips\n",
    "    node_x, node_y, hover_texts = [], [], []\n",
    "    node_colors, node_borders = [], []\n",
    "\n",
    "    for node in G.nodes():\n",
    "        x, y = pos[node]\n",
    "        node_x.append(x)\n",
    "        node_y.append(y)\n",
    "\n",
    "        # Get Genus attribute\n",
    "        genus = G.nodes[node].get(\"Genus\", \"Unclassified\")\n",
    "\n",
    "        # Assign colors: Red border for \"nan\", orange for \"Unclassified\", others get unique colors\n",
    "        if genus == \"nan\":\n",
    "            node_colors.append(\"red\")  # Node fill color\n",
    "            node_borders.append(\"red\")  # Node border color\n",
    "        elif genus == \"Unclassified\":\n",
    "            node_colors.append(\"gray\")  # Unclassified nodes\n",
    "            node_borders.append(\"black\")\n",
    "        else:\n",
    "            node_colors.append(color_map.get(genus, \"gray\"))  # Assign random colormap\n",
    "            node_borders.append(\"black\")\n",
    "\n",
    "        # Tooltip text\n",
    "        attrs = G.nodes[node]\n",
    "        tooltip = \"<br>\".join([f\"{key}: {value}\" for key, value in attrs.items()])\n",
    "        hover_texts.append(tooltip)\n",
    "\n",
    "    node_trace = go.Scatter(\n",
    "        x=node_x, y=node_y, mode='markers',\n",
    "        marker=dict(size=15, color=node_colors,\n",
    "                    line=dict(width=2, color=node_borders)),  # Border color\n",
    "        hoverinfo=\"text\", hovertext=hover_texts\n",
    "    )\n",
    "\n",
    "    fig = go.Figure(data=[edge_trace, node_trace],\n",
    "                    layout=go.Layout(\n",
    "                        title=\"HieVi\",\n",
    "                        showlegend=False, hovermode=\"closest\",\n",
    "                        margin=dict(b=0, t=50, l=0, r=0),\n",
    "                        xaxis=dict(showgrid=False, zeroline=False, showticklabels=False),\n",
    "                        yaxis=dict(showgrid=False, zeroline=False, showticklabels=False)\n",
    "                    ))\n",
    "\n",
    "    return fig  # Return the figure for further use"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 51,
   "metadata": {},
   "outputs": [
    {
     "ename": "ValueError",
     "evalue": "Mime type rendering requires nbformat>=4.2.0 but it is not installed",
     "output_type": "error",
     "traceback": [
      "\u001b[0;31m---------------------------------------------------------------------------\u001b[0m",
      "\u001b[0;31mValueError\u001b[0m                                Traceback (most recent call last)",
      "\u001b[0;32m/tmp/ipykernel_642016/3112704641.py\u001b[0m in \u001b[0;36m<module>\u001b[0;34m\u001b[0m\n\u001b[1;32m     10\u001b[0m \u001b[0;31m# Show the plot only if show=True\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m     11\u001b[0m \u001b[0;32mif\u001b[0m \u001b[0mshow\u001b[0m\u001b[0;34m:\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0;32m---> 12\u001b[0;31m     \u001b[0mfig\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mshow\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0m",
      "\u001b[0;32m~/anaconda3/envs/esmfold/lib/python3.7/site-packages/plotly/basedatatypes.py\u001b[0m in \u001b[0;36mshow\u001b[0;34m(self, *args, **kwargs)\u001b[0m\n\u001b[1;32m   3408\u001b[0m         \u001b[0;32mimport\u001b[0m \u001b[0mplotly\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mio\u001b[0m \u001b[0;32mas\u001b[0m \u001b[0mpio\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m   3409\u001b[0m \u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0;32m-> 3410\u001b[0;31m         \u001b[0;32mreturn\u001b[0m \u001b[0mpio\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mshow\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mself\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0;34m*\u001b[0m\u001b[0margs\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0;34m**\u001b[0m\u001b[0mkwargs\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0m\u001b[1;32m   3411\u001b[0m \u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m   3412\u001b[0m     \u001b[0;32mdef\u001b[0m \u001b[0mto_json\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mself\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0;34m*\u001b[0m\u001b[0margs\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0;34m**\u001b[0m\u001b[0mkwargs\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m:\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n",
      "\u001b[0;32m~/anaconda3/envs/esmfold/lib/python3.7/site-packages/plotly/io/_renderers.py\u001b[0m in \u001b[0;36mshow\u001b[0;34m(fig, renderer, validate, **kwargs)\u001b[0m\n\u001b[1;32m    393\u001b[0m         \u001b[0;32mif\u001b[0m \u001b[0;32mnot\u001b[0m \u001b[0mnbformat\u001b[0m \u001b[0;32mor\u001b[0m \u001b[0mVersion\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mnbformat\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0m__version__\u001b[0m\u001b[0;34m)\u001b[0m \u001b[0;34m<\u001b[0m \u001b[0mVersion\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0;34m\"4.2.0\"\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m:\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m    394\u001b[0m             raise ValueError(\n\u001b[0;32m--> 395\u001b[0;31m                 \u001b[0;34m\"Mime type rendering requires nbformat>=4.2.0 but it is not installed\"\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0m\u001b[1;32m    396\u001b[0m             )\n\u001b[1;32m    397\u001b[0m \u001b[0;34m\u001b[0m\u001b[0m\n",
      "\u001b[0;31mValueError\u001b[0m: Mime type rendering requires nbformat>=4.2.0 but it is not installed"
     ]
    }
   ],
   "source": [
    "#from utils.plotter import *\n",
    "# Plot the graph for small graphs only\n",
    "show = True # @param {type:\"boolean\"}\n",
    "if len(df) > 2048:\n",
    "    show = False\n",
    "fig = plot_hierarchical_graph(G,\"radial\")\n",
    "# Save to HTML\n",
    "fig.write_html(query_zarr_path[:-5] + \"_HieVi.html\")\n",
    "\n",
    "# Show the plot only if show=True\n",
    "if show:\n",
    "    fig.show()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 53,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Successfully created /media/microscopie-lcb/swapnesh/protein/embeddings/phages/1Sept2024_INPHARED_db_latest/subset.fasta with 114 sequences\n"
     ]
    }
   ],
   "source": [
    "def create_subset_fasta(input_fasta, output_fasta, accession_list):\n",
    "    \"\"\"\n",
    "    Create a new multifasta file containing only sequences matching the given accessions.\n",
    "    \n",
    "    Args:\n",
    "        input_fasta (str): Path to the input multifasta file\n",
    "        output_fasta (str): Path to the output multifasta file\n",
    "        accession_list (list): List of accession prefixes to include (e.g., ['phage-01', 'phage-02'])\n",
    "    \n",
    "    Returns:\n",
    "        bool: True if successful, False if there was an error\n",
    "    \"\"\"\n",
    "    try:\n",
    "        # Convert accession list to set for faster lookup\n",
    "        accession_set = set(accession_list)\n",
    "        current_header = None\n",
    "        current_sequence = []\n",
    "        sequences_to_keep = {}\n",
    "        \n",
    "        # Read the input file and collect matching sequences\n",
    "        with open(input_fasta, 'r') as f:\n",
    "            for line in f:\n",
    "                line = line.strip()\n",
    "                if line.startswith('>'):\n",
    "                    # Store previous sequence if it matches\n",
    "                    if current_header and current_sequence:\n",
    "                        full_sequence = ''.join(current_sequence)\n",
    "                        sequences_to_keep[current_header] = full_sequence\n",
    "                    \n",
    "                    # Start new sequence\n",
    "                    current_sequence = []\n",
    "                    current_header = line[1:]  # Remove '>'\n",
    "                    # Check if this header's accession is in our list\n",
    "                    accession = current_header.split('_')[0]\n",
    "                    if accession not in accession_set:\n",
    "                        current_header = None\n",
    "                elif current_header:\n",
    "                    current_sequence.append(line)\n",
    "            \n",
    "            # Store the last sequence if it matches\n",
    "            if current_header and current_sequence:\n",
    "                full_sequence = ''.join(current_sequence)\n",
    "                sequences_to_keep[current_header] = full_sequence\n",
    "        \n",
    "        # Write the matching sequences to output file\n",
    "        with open(output_fasta, 'w') as f:\n",
    "            for header, sequence in sequences_to_keep.items():\n",
    "                f.write(f\">new-{header}\\n\")\n",
    "                # Split sequence into lines of 60 characters (common FASTA format)\n",
    "                for i in range(0, len(sequence), 60):\n",
    "                    f.write(sequence[i:i+60] + \"\\n\")\n",
    "        \n",
    "        print(f\"Successfully created {output_fasta} with {len(sequences_to_keep)} sequences\")\n",
    "        return True\n",
    "        \n",
    "    except FileNotFoundError:\n",
    "        print(f\"Error: Input file '{input_fasta}' not found\")\n",
    "        return False\n",
    "    except PermissionError:\n",
    "        print(f\"Error: Cannot write to '{output_fasta}' - check permissions\")\n",
    "        return False\n",
    "    except Exception as e:\n",
    "        print(f\"Error: {str(e)}\")\n",
    "        return False\n",
    "\n",
    "# Sample parameters\n",
    "input_file = \"/media/microscopie-lcb/swapnesh/protein/embeddings/phages/1Sept2024_INPHARED_db_latest/1Sep2024_vConTACT2_proteins.faa\"\n",
    "output_file = \"/media/microscopie-lcb/swapnesh/protein/embeddings/phages/1Sept2024_INPHARED_db_latest/subset.fasta\"\n",
    "accessions = [\"OK499988\", \"OR756193\"]\n",
    "\n",
    "success = create_subset_fasta(input_file, output_file, accessions)\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "metadata": {},
   "outputs": [],
   "source": [
    "# required files\n",
    "'''\n",
    "Required files\n",
    "faiss_index for the right model\n",
    "the network gexf\n",
    "\n",
    "'''\n",
    "\n",
    "\n",
    "\n",
    "\n",
    "data_folder = \"/media/microscopie-lcb/swapnesh/protein/embeddings/phages/1Sept2024_INPHARED_db_latest/\"\n",
    "faiss_index_path = data_folder +\"ESM2_650m_1Sept24_650m.zarrfaiss_index.bin\"\n",
    "query_zarr_path = data_folder +\"ESM2_650m_1Sept24_650m.zarr\"#f\"{expt_name}_{model_name}.zarr\"\n",
    "db_clusters_path = data_folder +\"ESM2_3b_1Sept24_3b_flat_clusters_3b.csv\"\n",
    "eps_values_path = data_folder +\"ESM2_650m_1Sept24_650m_eps_values_flat_clusters.npy\"\n",
    "db_zarr_path = data_folder +\"ESM2_650m_1Sept24_650m.zarr\"\n",
    "HieVi_INPHARED_ordered_annotation = \"HieVi_INPHARED_ordered_annotation.csv\"\n",
    "HieVi_tree = data_folder + \"ESM2_3b_1Sept24_3b.zarr.gexf\"\n",
    "\n",
    "htree = nx.read_gexf(HieVi_tree)\n",
    "\n",
    "zarr_store = zarr.open(db_zarr_path,'r')\n",
    "db_accessions = zarr_store['accessions'][:]\n",
    "\n",
    "annotations = pd.read_csv(HieVi_INPHARED_ordered_annotation)\n",
    "annotations= annotations[annotations[\"Accession\"].isin(db_accessions)]\n",
    "annotations = annotations.set_index(\"Accession\").loc[db_accessions].reset_index()\n",
    "\n",
    "eps_values = np.load(eps_values_path)\n",
    "\n",
    "\n",
    "index = faiss.read_index(faiss_index_path)\n",
    "\n",
    "hievi_cluster = pd.read_csv(db_clusters_path)\n",
    "hievi_cluster= hievi_cluster[hievi_cluster[\"Accession\"].isin(db_accessions)]\n",
    "hievi_cluster = hievi_cluster.set_index(\"Accession\").loc[db_accessions].reset_index()\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "metadata": {},
   "outputs": [],
   "source": [
    "filename = \"/media/microscopie-lcb/swapnesh/protein/embeddings/phages/NewGenomes/HieVi_test_case.faa\"\n",
    "# @title Compute phage representations\n",
    "expt_name = \"MyPhages\"  # @param {type:\"string\"}\n",
    "output_folder = os.path.dirname(filename) + os.sep \n",
    "fasta_path = filename\n",
    "model_name = \"650m\" # This colab works for 650m only\n",
    "mode = \"mean\"\n",
    "#!python GenPhageRepresentationsESM2.py {expt_name} {output_folder} {fasta_path} {model_name} {mode}\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "metadata": {},
   "outputs": [],
   "source": [
    "query_zarr_path = f\"{output_folder}{expt_name}_{model_name}.zarr\"\n",
    "\n",
    "query_zarr_store = zarr.open(query_zarr_path,'r')\n",
    "query_vectors = query_zarr_store['vectors_mean'][:]*1.0\n",
    "query_accessions = query_zarr_store['accessions'][:]\n",
    "\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "(54, 1280)"
      ]
     },
     "execution_count": 39,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": 38,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "array([ 0.00258736, -0.00066257,  0.00058419, ..., -0.01147296,\n",
       "       -0.00221943,  0.00873887], dtype=float32)"
      ]
     },
     "execution_count": 38,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "hievi_cluster_prefix = 'HC_p'\n",
    "k_neighbours = 64 # \n",
    "distance_threshold = 0.023 # \n",
    "\n",
    "distances, indices = index.search(query_vectors, 1)\n",
    "valid_idx = distances[:,0] < distance_threshold\n",
    "invalid_idx = np.logical_not(valid_idx)\n",
    "\n",
    "if len(np.where(invalid_idx)[0]):\n",
    "    print('Cannot classifiy: ' ,len(np.where(invalid_idx)[0]))\n",
    "    invalid_query_df = pd.DataFrame({\"Accession\": query_accessions[invalid_idx]})\n",
    "    invalid_query_df.to_csv(query_zarr_path[:-5] + \"_HieVi_Unclassifieds.csv\")\n",
    "\n",
    "D = np.squeeze(distances[valid_idx,0])\n",
    "I = np.squeeze(indices[valid_idx,0])\n",
    "\n",
    "D1, I1 = index.search(np.array(zarr_store['vectors_mean'])[I]*1.0, k_neighbours)\n",
    "\n",
    "all_indices = []\n",
    "for ds,idxs in zip(I1,D1):\n",
    "    for d,i in zip(ds,idxs):\n",
    "        if d < distance_threshold*10:\n",
    "            all_indices +=[i]\n",
    "\n",
    "\n",
    "all_indices = np.unique(np.array(all_indices+ list(I)))\n",
    "all_nearest_accessions = db_accessions[all_indices]\n",
    "\n",
    "new_accession = []\n",
    "for a in all_nearest_accessions:\n",
    "    new_accession += find_predecessor_and_leaves(htree, a,2)\n",
    "    \n",
    "\n",
    "all_nearest_accessions = np.unique(np.ravel(np.array(new_accession)))\n",
    "all_indices = np.array([np.where(db_accessions==a)[0] for a in all_nearest_accessions])\n",
    "\n",
    "subset_db_vectors = np.squeeze(np.array(zarr_store['vectors_mean'])[all_indices]*1.0)\n",
    "subset_db_vectors.shape\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 40,
   "metadata": {},
   "outputs": [],
   "source": [
    "# @title Combine query and nearest phages from database data\n",
    "\n",
    "# indices = np.unique(np.ravel(indices))\n",
    "# print(f\"Nearest neighbor search completed. Found {len(indices)} unique neighbors.\")\n",
    "# Get nearest accessions in tree\n",
    "nearest_accessions = annotations[annotations[\"Accession\"].isin(all_nearest_accessions)]\n",
    "nearest_accessions = nearest_accessions.set_index(\"Accession\").loc[all_nearest_accessions].reset_index()\n",
    "\n",
    "\n",
    "# Combine query and database data\n",
    "query_df = pd.DataFrame({\"Accession\": query_accessions[valid_idx]})\n",
    "annotation_df = pd.concat([nearest_accessions, query_df], axis=0)\n",
    "mprs = np.concatenate((subset_db_vectors, query_vectors[valid_idx]), axis=0)\n",
    "\n",
    "# Perform clustering\n",
    "# dist_scaled = euclidean_distances(mprs).astype(\"double\")\n",
    "# clusterer = hdbscan.HDBSCAN(\n",
    "#     min_cluster_size=2,\n",
    "#     n_jobs=32,\n",
    "#     min_samples=1,\n",
    "#     allow_single_cluster=False,\n",
    "#     cluster_selection_method=\"leaf\",\n",
    "#     metric=\"precomputed\",\n",
    "#     gen_min_span_tree=True\n",
    "# )\n",
    "# clusterer.fit(dist_scaled)\n",
    "\n",
    "clusterer = hdbscan.HDBSCAN(\n",
    "    min_cluster_size=2,\n",
    "    min_samples=1,\n",
    "    allow_single_cluster=False,\n",
    "    cluster_selection_method=\"leaf\",\n",
    "    metric=\"euclidean\",\n",
    "    gen_min_span_tree=True\n",
    ")\n",
    "clusterer.fit(mprs)\n",
    "\n",
    "annotation_df[\"HieVi_granular_cluster\"] = clusterer.labels_\n",
    "for i,eps in enumerate(eps_values):\n",
    "    annotation_df[hievi_cluster_prefix+str(i)] = clusterer.dbscan_clustering(cut_distance=eps,min_cluster_size=2)\n",
    "#annotation_df"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 41,
   "metadata": {},
   "outputs": [],
   "source": [
    "# @title Generate Network\n",
    "\n",
    "\n",
    "min_lambda =13.00 # @param {type:\"slider\", min:-1, max:32, step:1}\n",
    "# Create and save network\n",
    "G = make_network(clusterer, annotation_df,min_lambda=min_lambda)\n",
    "nx.write_gexf(G, query_zarr_path[:-5] + \"_HieVi.gexf\")\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "esmfold",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.7.12"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
